### PR TITLE
Fix SetupArgoCD function and improve its test coverage

### PR DESCRIPTION
**Pull request by jopit**
#### Description:
- Updated the `SetupArgoCD` function to ensure if the resources it is trying to create/update already exist, it will GET the resource before doing an UPDATE
- Added unit tests
- Added an e2e test

### CODE REVIEW

File: cluster-agent/utils/argocd_create.go [hunk 1]
```diff
@@ -310,22 +310,28 @@ func SetupArgoCD(ctx context.Context, apiHost string, argoCDNamespace string, k8
...
-			logutil.LogAPIResourceChangeEvent(serviceAccount.Namespace, serviceAccount.Name, serviceAccount, logutil.ResourceModified, log)
-
-		} else {
-			return fmt.Errorf("error on Create %v", err)
+	actualServiceAccount := &corev1.ServiceAccount{}
+
+	err := k8sClient.Get(ctx, client.ObjectKeyFromObject(serviceAccount), actualServiceAccount)
+	if err == nil {
+		// It already exists, so we update it
+		actualServiceAccount.Secrets = serviceAccount.Secrets
```
Comment (keithchong) [inline cluster-agent/utils/argocd_create.go:318]: Do these do anything?

File: cluster-agent/utils/argocd_create.go [hunk 2]
```diff
@@ -337,20 +343,31 @@ func SetupArgoCD(ctx context.Context, apiHost string, argoCDNamespace string, k8
...
+			"kubernetes.io/service-account.name": serviceAccount.Name,
+		}
+		actualSecret.Type = "kubernetes.io/service-account-token"
+		if err := k8sClient.Update(ctx, actualSecret); err != nil {
+			return fmt.Errorf("error on Update secret: %w", err)
 		}
+		logutil.LogAPIResourceChangeEvent(actualSecret.Namespace, actualSecret.Name, actualSecret, logutil.ResourceModified, log)
+	} else if apierr.IsNotFound(err) {
+		// It doesn't exist, so we create it
+		if err := k8sClient.Create(ctx, secret); err != nil {
```
Comment (keithchong) [inline cluster-agent/utils/argocd_create.go:363]: Question: Do we want to add the annotation as above?

File: cluster-agent/utils/argocd_create.go [hunk 3]
```diff
@@ -366,20 +383,27 @@ func SetupArgoCD(ctx context.Context, apiHost string, argoCDNamespace string, k8
...
 		}
+		logutil.LogAPIResourceChangeEvent(actualClusterRole.Namespace, actualClusterRole.Name, actualClusterRole, logutil.ResourceModified, log)
+	} else if apierr.IsNotFound(err) {
+		// It doesn't exist, so we create it
+		if err := k8sClient.Create(ctx, &clusterRole); err != nil {
+			return fmt.Errorf("error on Create cluster role: %w", err)
+		}
+		logutil.LogAPIResourceChangeEvent(clusterRole.Namespace, clusterRole.Name, clusterRole, logutil.ResourceCreated, log)
+	} else {
+		// Some other error occured
```
Comment (keithchong) [inline cluster-agent/utils/argocd_create.go:404]: occurred

File: cluster-agent/utils/argocd_create.go [hunk 4]
```diff
@@ -366,20 +383,27 @@ func SetupArgoCD(ctx context.Context, apiHost string, argoCDNamespace string, k8
...
-			}
-			logutil.LogAPIResourceChangeEvent(clusterRole.Namespace, clusterRole.Name, clusterRole, logutil.ResourceModified, log)
-
-		} else {
-			return fmt.Errorf("error on Create %v", err)
+	actualClusterRole := rbac.ClusterRole{}
+
+	err = k8sClient.Get(ctx, client.ObjectKeyFromObject(&clusterRole), &actualClusterRole)
+	if err == nil {
+		// It already exists, so we update it.
```
Comment (keithchong) [inline cluster-agent/utils/argocd_create.go:390]: Were you intending to add a unit test in argocd_create_test.go to test this err == nil if condition?  I only see the clusterRoler being created.

File: cluster-agent/utils/argocd_create.go [hunk 5]
```diff
@@ -399,23 +423,30 @@ func SetupArgoCD(ctx context.Context, apiHost string, argoCDNamespace string, k8
...
-			logutil.LogAPIResourceChangeEvent(clusterRoleBinding.Namespace, clusterRoleBinding.Name, clusterRoleBinding, logutil.ResourceModified, log)
-
-		} else {
-			return fmt.Errorf("error on Create %v", err)
+	actualClusterRoleBinding := rbac.ClusterRoleBinding{}
+
+	err = k8sClient.Get(ctx, client.ObjectKeyFromObject(&clusterRoleBinding), &actualClusterRoleBinding)
+	if err == nil {
+		// It already exists, so we update it.
+		actualClusterRoleBinding.RoleRef = clusterRoleBinding.RoleRef
```
Comment (keithchong) [inline cluster-agent/utils/argocd_create.go:431]: Same with this, I don't see a unit test for this if block

File: cluster-agent/utils/argocd_create_test.go [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+					Type:      corev1.SecretType("example"),
+				}
+				err = k8sClient.Create(ctx, clusterSecret)
+				Expect(err).ToNot(HaveOccurred())
+
+				By("calling SetupArgoCD()")
+				err = SetupArgoCD(ctx, apiHost, namespace.Name, k8sClient, logger)
+				Expect(err).ToNot(HaveOccurred())
+
+				By("ensuring the sevice account was updated")
```
Comment (keithchong) [inline cluster-agent/utils/argocd_create_test.go:251]: service

File: cluster-agent/utils/argocd_create_test.go [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+				}))
+			})
+		})
+
+		When("SetupArgoCD() is called and the resources already exist", func() {
+			It("creates the required service account, secret, cluster role, rolebinding and cluster secret", func() {
+				By("starting a goroutine to add a token to the secret created by SetupArogCD()")
+				startSecretUpdater(ctx, k8sClient)
+
+				By("creating the sevice account")
```
Comment (keithchong) [inline cluster-agent/utils/argocd_create_test.go:155]: service

File: cluster-agent/utils/argocd_create_test.go [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+			Expect(err).ToNot(HaveOccurred())
+			Expect(namespace).ToNot(BeNil())
+
+			k8sClient = fake.NewClientBuilder().WithScheme(scheme).WithObjects(namespace).Build()
+
+		})
+
+		When("SetupArgoCD() is called", func() {
+			It("creates the required service account, secret, cluster role, rolebinding and cluster secret", func() {
+				By("starting a goroutine to add a token to the secret created by SetupArogCD()")
```
Comment (keithchong) [inline cluster-agent/utils/argocd_create_test.go:56]: SetupArgoCD

File: cluster-agent/utils/argocd_create_test.go [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+				startSecretUpdater(ctx, k8sClient)
+
+				By("calling SetupArgoCD()")
+				err := SetupArgoCD(ctx, apiHost, namespace.Name, k8sClient, logger)
+				Expect(err).ToNot(HaveOccurred())
+
+				By("ensuring the sevice account was created")
```
Comment (keithchong) [inline cluster-agent/utils/argocd_create_test.go:63]: service

File: cluster-agent/utils/argocd_create_test.go [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+					"name":   ClusterSecretName,
+					"server": apiHost,
+					"config": string(jsonString),
+				}))
+			})
+		})
+
+		When("SetupArgoCD() is called and the resources already exist", func() {
+			It("creates the required service account, secret, cluster role, rolebinding and cluster secret", func() {
+				By("starting a goroutine to add a token to the secret created by SetupArogCD()")
```
Comment (keithchong) [inline cluster-agent/utils/argocd_create_test.go:152]: SetupArgoCD

File: tests-e2e/argocd/argocd_instance_test.go [hunk 11]
```diff
@@ -126,5 +126,93 @@ var _ = Describe("Standalone ArgoCD instance E2E tests", func() {
...
+			}
+
+			Eventually(argocdInstance, "60s", "5s").Should(k8s.ExistByName(k8sClient))
+			Expect(err).ToNot(HaveOccurred())
+
+			By("ensuring ArgoCD resource exists in kube-system namespace")
+			err = argocdv1.SetupArgoCD(ctx, apiHost, argocdNamespace, k8sClient, log)
+			Expect(err).ToNot(HaveOccurred())
+
+			By("ensuring the SetupArgoCD function works even if resources already exist in kube-system namespace")
```
Comment (keithchong) [inline tests-e2e/argocd/argocd_instance_test.go:158]: Is it possible to refactor and maybe common up this test case with at least the previous one?  From what I see this block of code is the only difference.

File: cluster-agent/utils/argocd_create.go (hunk 4 above)
Comment (jopit) [reply cluster-agent/utils/argocd_create.go:390]: Good catch, I've updated the unit test.

File: cluster-agent/utils/argocd_create_test.go (hunk 8 above)
Comment (jopit) [reply cluster-agent/utils/argocd_create_test.go:56]: Fixed

File: cluster-agent/utils/argocd_create_test.go (hunk 9 above)
Comment (jopit) [reply cluster-agent/utils/argocd_create_test.go:63]: Fixed

File: cluster-agent/utils/argocd_create_test.go (hunk 7 above)
Comment (jopit) [reply cluster-agent/utils/argocd_create_test.go:155]: Fixed

File: cluster-agent/utils/argocd_create_test.go (hunk 6 above)
Comment (jopit) [reply cluster-agent/utils/argocd_create_test.go:251]: Fixed

File: tests-e2e/argocd/argocd_instance_test.go (hunk 11 above)
Comment (jopit) [reply tests-e2e/argocd/argocd_instance_test.go]: Fixed

**Review by jgwest — APPROVED** [review]
LGTM, thanks @jopit, and thanks @keithchong for reviewing!

**Pull request merged**